lib/ecto/changeset.ex

```diff
@@ -209,6 +242,23 @@ defmodule Ecto.Changeset do
     end
   end
 
+  defp merge_changesets(%Ecto.Changeset{model: model} = cs1, %Ecto.Changeset{model: model} = cs2) do
+    new_params   = Map.merge(cs1.params, cs2.params)
+    new_changes  = Map.merge(cs1.changes, cs2.changes)
+    new_errors   = cs1.errors ++ cs2.errors
+    new_required = Enum.uniq(cs1.required ++ cs2.required)
+
+    # Merges the :optional field of both changesets and then pulls out all the
+    # fields that have now become required.
+    opt_set = Enum.into(cs1.optional ++ cs2.optional, HashSet.new)
```

Comment:
I think this will be more performant (and clearer) if you simply write it as `Enum.uniq(optional1 ++ optional2) -- required` :)